File: recursion/capitalize_words.py
# Write a recursive function called capitalize_words. Given a list of words, return a new list containing each word capitalized.
def capitalize_words(string_list):
    # Base case
    if len(string_list) == 1:
        return [string_list[0].upper()]
    elif len(string_list) == 0:
        return []

    # Recursion
    return [string_list[0].upper(), *capitalize_words(string_list[1:])]

# Tests
print(capitalize_words([]) == [])
print(capitalize_words(['what', 'is', 'recursion']) == ['WHAT', 'IS', 'RECURSION'])
print(capitalize_words(['it', 'is']) == ['IT', 'IS'])
print(capitalize_words(['recursion']) == ['RECURSION'])

File: recursion/capitalizeWords.ts
// Write a recursive function called capitalizeWords. Given an array of words, return a new array containing each word capitalized.
import _ from 'lodash';

function capitalizeWords(arr: string[]): string[] {
	// Base cases
	if (arr.length === 1) return [(arr[0] ?? '').toUpperCase()];
	if (arr.length === 0) return [];

	// Recursion
	const currentStr: string = arr[0] ?? '';
	const newStr: string = currentStr.toUpperCase();
	return [newStr].concat(capitalizeWords(arr.slice(1)));
}

// Tests
console.log(
	_.isEqual(capitalizeWords(['what', 'is', 'recursion']), [
		'WHAT',
		'IS',
		'RECURSION',
	])
);
console.log(_.isEqual(capitalizeWords([]), []));
